agent/internal/containers.go

```diff
@@ -144,32 +145,43 @@ func (c *containerManager) overwriteSpec(
 		slotIds = append(slotIds, strconv.Itoa(d.ID))
 	}
 	spec.RunSpec.ContainerConfig.Labels[dockerContainerDevicesLabel] = strings.Join(slotIds, ",")
-	if len(cont.Devices) > 0 && cont.Devices[0].Type == device.GPU {
-		spec.RunSpec.HostConfig.Runtime = nvidiaRuntime
+
+	spec = c.overwriteSpecforGPUs(cont, spec)
+
+	return spec
+}
+
+func (c *containerManager) overwriteSpecforGPUs(
+	cont cproto.Container, spec container.Spec,
+) container.Spec {
+	useGPU := len(cont.Devices) > 0 && cont.Devices[0].Type == device.GPU
+	spec.RunSpec.ContainerConfig.Env = append(
+		spec.RunSpec.ContainerConfig.Env,
+		fmt.Sprintf("DET_USE_GPU=%t", useGPU),
+	)
+	if useGPU {
+		var uuids []string
+		for _, d := range cont.Devices {
+			uuids = append(uuids, d.UUID)
+		}
+		request := dcontainer.DeviceRequest{
+			Driver:       "nvidia",
+			Capabilities: [][]string{{"gpu", "utility"}},
+			DeviceIDs:    uuids,
+		}
+		spec.RunSpec.HostConfig.DeviceRequests = append(spec.RunSpec.HostConfig.DeviceRequests, request)
 	}
 	return spec
 }
 
 func (c *containerManager) containerEnvVars(cont cproto.Container) []string {
-	visibleDevices := "none"
-	useGPU := false
 	var slotIds []string
-	var slotUUIDs []string
 	for _, d := range cont.Devices {
 		slotIds = append(slotIds, strconv.Itoa(d.ID))
-
-		if useGPU = d.Type == device.GPU; useGPU {
-			slotUUIDs = append(slotUUIDs, d.UUID)
-		}
-	}
-	if len(slotUUIDs) > 0 {
-		visibleDevices = strings.Join(slotUUIDs, ",")
 	}
 	return []string{
 		fmt.Sprintf("DET_CONTAINER_ID=%s", cont.ID),
-		fmt.Sprintf("DET_USE_GPU=%t", useGPU),
 		fmt.Sprintf("DET_SLOT_IDS=[%s]", strings.Join(slotIds, ",")),
-		fmt.Sprintf("NVIDIA_VISIBLE_DEVICES=%s", visibleDevices),
```

Comment:
Yes, but I don't think it's that problematic. We already required Docker 19.03 (without, I think, realizing it, but we did). Installing `nvidia-docker2` (the old thing that adds a runtime) has also installed `nvidia-container-toolkit` (the new thing being switched to here) since at most three days after the release of Docker 19.03 ([2019-07-22](https://docs.docker.com/engine/release-notes/#19030) to [2019-07-25](https://github.com/NVIDIA/nvidia-docker/commit/36e7a11c52f300668099ee500791e07302810007)). I may have missed something, but, as far as I know, there is essentially no scenario in which someone can run the agent but be more than a system package upgrade away from not requiring the agent to support the runtime (and even that only if they haven't upgraded in the past nine months).

Reply:
In that case can you also run the e2e-gpu tests (this can be done by pushing an upstream branch and manually approving the run GPU test job). I want to make sure that this doesn't break our existing environments. Also it's probably worth bringing in @seanr15 so that we can verify that this won't break Recursion. I think it's probably worth understanding if we're going to have to ask them to upgrade all of their on-prem machines or not.

Reply:
In that case can you also run the e2e-gpu tests (this can be done by pushing an upstream branch and manually approving the run GPU test job). I want to make sure that this doesn't break our existing environments. Also it's probably worth bringing in @seanr15 so that we can verify that this won't break existing customers. I think it's probably worth understanding if we're going to have to ask anyone to upgrade their on-prem machines or not.

Reply:
GPU tests have passed: https://app.circleci.com/pipelines/github/determined-ai/determined/979/workflows/f5570946-4436-4f61-8e15-1943f8c74633.